### ISSUE TITLE
Install loops forever, no easy way to trace install progress of diskful nodes.

### ISSUE THREAD

**Issue by whowutwut** [xcat2/xcat-core#4582]
There's no way to easily trace the process of installing a compute node in the xCAT log to determine what causes a compute node to fail getting the chain properties updated correctly and stop from going in an infinite loop.   (Essentially `#boot` needs to be set in the petitboot config file) 

In some of my debug, I came up with the following high level trace of what is going on...

### post install processing 

The interesting script is post.xcat , inside here:
   1.  downloads postscripts
   2.  executes mypostscript.post
   3.  Some point, we run nodeset <> next
   4. This triggers petitboot.pm to run rsetboot <> default (This seem to cause some failure and resulted in the rest of the script not executing, did it kill monitor process?)
   5. Then runs makedhcp
   6. Then runs setstate()
   7.  Inside setstate() it takes the chain.currentstate and either writes #boot into petitboot config file or generates the kickstart file again

If something here fails, inside post processing, it seems like we will be in an infinite install loop.

### Debug install logs 

Target node: mid05tor12cn05 

Started install diskful here:
```
====================================================
[Date]       2017-12-21 13:46:03
[ClientType] cli
[Request]    rpower mid05tor12cn05 reset
[Response]
```

Attaching the full `cluster.log` from provisioning: 
[mid05tor12cn05.cluster.log](https://github.com/xcat2/xcat-core/files/1580369/mid05tor12cn05.cluster.log)

If we do the following command:  `grep mid05tor12cn05 /var/log/xcat/cluster.log`

it doesn't contain things like: 
```
[root@briggs01 xcat]# grep next mid05tor12cn05.cluster.log
Dec 21 13:59:09 briggs01 xcat[11497]: DEBUG xcatd: dispatch request 'nodeset next' to plugin 'petitboot'
Dec 21 13:59:09 briggs01 xcat[11497]: DEBUG xcatd: dispatch request 'runbeginpre next' to plugin 'prescripts'
Dec 21 13:59:09 briggs01 xcat[11497]: DEBUG xcatd: dispatch request 'setdestiny next' to plugin 'destiny'
Dec 21 13:59:10 briggs01 xcat[11497]: DEBUG xcatd: dispatch request 'runendpre next' to plugin 'prescripts'
```

So how can we get a full picture of what is going on, how do we see when problems happen so Admins know how to resolve this? 

Logs need to have consistent searchable keywords
```
[root@briggs01 xcat]# grep -i error mid05tor12cn05.cluster.log
[root@briggs01 xcat]# grep -i failed mid05tor12cn05.cluster.log
Dec 21 13:59:07 mid05tor12cn05 xcat:  failed to download precreated mypostscript, trying to generate with getpostscript.awk
[root@briggs01 xcat]#
```

This is a success case, I don't know how we could inject errors along the way to see if we could actually figure out why the chain table does not get updated..

**Comment by robin2008** [comment]
Yes, there are some gaps for logs (cluster.log only for the logs from xcatd,  command.log for each command result, and compute log for the logs  from compute nodes syslog)

So sometime we need to concentrate those logs together for debugging an provisiong issue, and I think `xcatprobe osdeploy -r` (Replay history) is designed for this purpose.

However, it cannot output the concentrated logs even if it does already gathered them (to set debug flag in  /opt/xcat/probe/subcmds/osdeploy could show the raw information).

In addition,  just as you point out,  it is hard to parse the cluster.log to create the association with compute.log.  As `cluster.log` does not log every important things, especially for node level, so the log source is not enough.  I think the origin purpose is considering not too many logs to impact admin as xCAT usually do operations on large amount of nodes.

I also would like to make the logging be enhanced for better using, especially for provisioning process, hw control, updatenode.  I think we can consider part of them in 2.13.10

**Comment by whowutwut** [comment]
@zet809 @robin2008 It seems like recycling xCAT on the MN resolves this issue... We need to look into this for next year at high priority.

**Comment by robin2008** [comment]
xcatd will fork  some child process there.
```
# ps -ef|grep xcatd:|grep -v grep
root      11493      1  0 Dec20 ?        00:00:01 xcatd: SSL listener
root      11494  11493  0 Dec20 ?        00:01:51 xcatd: DB Access
root      11496  11493  0 Dec20 ?        00:00:47 xcatd: UDP listener
root      11497  11493  0 Dec20 ?        00:00:00 xcatd: install monitor
root      11498  11496  0 Dec20 ?        00:00:00 xcatd: Discovery worker
root      11499  11493  0 Dec20 ?        00:00:00 xcatd: Command log writer
```
If some error cause some of them quit, xcatd may not work as normal again before restarting.
I guess the issue you hit might be related to `install monitor` is died due to some unexpected error.

So one is to do more strict error handling the child process, especially for those we often see the issue. It can reduce the possibility into such the situation 

Another, we could let the  main process monitor its childer, and restart them when panic. It could self-heal when unexpected things.

Third, in xCAT, it has some process-level cache which means the cache only exists in process life-cycle (xCAT plugins run as a child process, so plug-ins would be benefited from the cache )
But for such long-running child process, it will get the negative impacted if the cache is not refreshed in time (normally those process-level cache will not refresh). For example, it cache the name and IP, after the IP is changed, xcatd will always get the old IP.

**Comment by robin2008** [comment]
We have made some log enhancement in https://github.com/xcat2/xcat-core/pull/4648
And for the logs in xcatd plugins or postscripts, it is still required to be enhanced case by case.

In addition, now as the dhcpd and httpd/tftpd is not controlled by xCAT, we don't have their logs in cluster.log. Perhaps xcatd could monitor their logs to grab the important logs into cluster.log, then we can have a full picture of provisioning. (Discuss it later)

**Comment by dombrowa** [comment]
As I just came across this problem on the firestone cluster in Yorktown installing an Ubuntu cloud image: I have noticed that the information of /var/log/xcat/xcat.log on the provisioned node does not get posted on the management node nor is there any good trace in rcons to see what is going on. The installed shows an unreliable progress bar even when the node is already rebooting. To tap into the install loop I did a brute force approach:
At first, I monitored via nodestat where the node is and when I can ssh back into it before it loops again
Second I ran a watch command to rsync the nodes /var/log/xcat/xcat.log command to see if anything during the install failed (which  by now I learned was not causing the install loop)
In any case, I which to see a feature of retrieving more info during the node install on or from the management node. Would this be a reasonable request to xCAT ?

**Comment by robin2008** [comment]
@dombrowa Do you have try the debug mode? To set `xcatdebugmode=1` and then more logs during the installation will be logged to xCAT MN via syslog,  you can check them by `tail -f computes.log | grep <mycn>`

**Comment by whowutwut** [comment]
@dombrowa  when this happens, can you quickly run `xcatprobe xcatmn` and just record this output, I think it's the monitor daemon that goes down causing the updateflag command to fail and so this does not trigger changes to the petitboot config file. 

@robin2008  I think we have to do a little more on the debug for this, if it's a critical error, we should not have to turn `xcatdebugmode=1` to gather more info, something should be logged automatically.

**Comment by dombrowa** [comment]
I have not tried xcatdebugmode=1 and have not gathered further info after the node installed properly. You may close this issue for reference.

**Comment by robin2008** [comment]
@whowutwut Already adds more logs in non-debug mode, but still only important logs will reported to xCAT mn. For release prior to 2.13.10, using debug mode is still the only way to get more information.

To use debug mode,  the syslog forwarding will be enabled in the native OS installer `anaconda/preceed`, there would be lots of syslogs, so it is only recommended for a debug purpose when hit problem.

**Comment by immarvin** [comment]
@whowutwut @robin2008 , can this ticket be closed now?

**Comment by whowutwut** [comment]
Ok, let's close for now and see if we new issues arise.

**Comment by robin2008** [comment]
Let's close it first.
- Now `install monitor` will not exit in most of the case, and we can have the logs there to know it is quit.
- xCAT has both the console log (consoleondemand=on in site table) method and syslog method to fetch the logs from compute nodes (/var/log/xcat/computes.log)

**Comment by robin2008** [comment]
reopen it to cover all die cases in `install monitor`.

There are three major task in `install monitor` for a node diskful installation when it request boot from local disk before reboot:

- install monitor could accept the updateflag requesting (depends: alive and listening)
- Match source IP address of this request with a defined node (depends: node must send the updateflag using the provision IP)
- run nodeset ... next to modify the petitiboot/grub2/xnba config file successfully. (depends: plugin code logic)

Now the part 3 is already covered in `eval` block, if something die in plugins, we can see the logs like "possible BUG encountered by xCAT install monitor". And it won't cause the install monitor exit as we catch the error and continue with logging error message.

so in this ticket, we plan to add the eval for step1 and step2, this cause install monitor exit without any error messages.

**Comment by robin2008** [comment]
Address it in https://github.com/xcat2/xcat-core/pull/5005

**Comment by robin2008** [comment]
close for now and see if we new issues arise.
